daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala

```diff
@@ -67,24 +95,155 @@ class EvaluationOrderTest extends AnyWordSpec with Matchers {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = False, key = M:toKey sig, nested = M:buildNested 0 }
+           """,
+          Array(SParty(alice), SParty(bob)),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[TemplatePreconditionViolated]
+
+        }
+        msgs shouldBe Seq(
+          "precondition"
+        )
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with duplicate contract key
+      "duplicate contract key" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) ->
+                let c: M:T = M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 }
+                in sbind x : ContractId M:T <- create @M:T c
+                  in create @M:T c
+           """,
+          Array(SParty(alice), SParty(bob)),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[DuplicateContractKey]
+        }
+        msgs shouldBe allTraces ++ allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with empty contract key maintainers
+      "empty contract key maintainers" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:keyNoMaintainers, nested = M:buildNested 0 }
+           """,
+          Array(SParty(alice), SParty(bob)),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[CreateEmptyContractKeyMaintainers]
+
+        }
+        msgs shouldBe allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with authorization failure
+      "authorization failure" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 }
+           """,
+          Array(SParty(alice), SParty(bob)),
+          bob,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[FailedAuthorization]
+
+        }
+        msgs shouldBe allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with contract id in contract key
+      "contract id in contract key" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) (cid : ContractId Unit) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:toKeyWithCid sig cid, nested = M:buildNested 0 }
```

Comment:
Shouldn't this (and the above failure cases) have `M:buildNested 100` to make sure the nesting error does not pre-empt the other errors?

Reply:
As mentioned in the PR description and in the comment, I’m excluding ordering of uncatchable errors relative to each other. If we mess that up, we change error messages which is annoying but not all that problematic. Catchable vs non-cachable or other catchable errors on the other hand breaks replays after updates which can brick ledgers so I’d like to focus on that. Not opposed to extending it later though to also test the other cases.

Reply:
Ah, ok, thanks for the explanation.

The nesting error could still pre-empt a catcheable error (exceptions being thrown), but probably the check for that should go in the SemanticsEvalOrder tests. (You're kind of ruling it out here with the traces, in any case.)